src/produit.service.ts

```diff
@@ -0,0 +1,322 @@
+import { Http, Headers, RequestOptions } from '@angular/http';
+import { Injectable } from '@angular/core';
+import { Observable, BehaviorSubject, ReplaySubject } from 'rxjs/Rx';
+import { config } from './config';
+
+@Injectable()
+export class ProductService {
+
+    public product$: ReplaySubject<Product>;
+    private headers: Headers;
+    config: any;
+    private cache: Cache = new Cache();
+
+    constructor(
+        private http: Http
+    ) {
+        this.config = config;
+        this.headers = new Headers();
+        this.headers.append('Content-Type', 'application/json');
+
+        //
+        // 1 means to keep the last value
+        this.product$ = new ReplaySubject(1);
+    }
+
+    private deleteCache(product: Product) {
+        if (this.cache.map[product.sku]) {
+            this.cache.map.delete(product.sku);
+            let index = this.cache.list.indexOf(product)
+            if (index > -1)
+                this.cache.list.splice(index, 1);
+        }
+    }
+
+    private updateCache(product: Product) {
+        //
+        //check if already exist on cache and add in it if not the case
+        if (!this.cache.map[product.sku]) {
+            this.cache.map[product.sku] = product;
+            this.cache.list.push(product);
+            return product;
+        }
+        //update existing entry
+        return Object.assign(this.cache.map[product.sku], product);
+    }
+
+
+    //
+    // REST api wrapper
+    //
+
+    select(filter?: any): Observable<Product[]> {
+        filter = filter || {};
+        return this.http.get(this.config.API_SERVER + '/v1/products', {
+            search: filter,
+            headers: this.headers,
+            withCredentials: true
+        })
+            .map(res => res.json() as Product[])
+            // TODO manage cache!
+            .map(product => product.map(this.updateCache.bind(this)))
+            .catch(this.handleError);
+    };
+
+    findByLocationCategoryAndDetail(category, detail): Observable<Product[]> {
+        return this.http.get(this.config.API_SERVER + '/v1/products/location/' + location + '/category/' + category + '/details/' + detail, {
+            headers: this.headers,
+            withCredentials: true
+        }).map(res => res.json() as Product[]);
+        // TODO manage cache!
```

Comment:
pourquoi tu n'ajoute pas `.map(products => product.map(this.updateCache.bind(this)))` ici et aux suivantes?

Reply:
il faut aussi remplacer partout: `  .map(res => res.json().map(obj => new Product(obj)));` attention Product doit modifier son constructeur